File: python/java_to_python/06-pgmAvion/src/co/edu/campusucc/poo/Avion.py
from abc import ABC, abstractmethod
import time

class Avion(ABC):
    def __init__(self):
        self.marca = None
        self.tipo = None
        self.velocidad = 0

    def __str__(self):
        return "Clase Abstracta--> Avion: " + self.marca + " " + self.tipo + " "

    # Abstract Behavior
    @abstractmethod
    def startingEngine(self):
        pass

    @abstractmethod
    def stopingEngine(self):
        pass

    # Definir comportamiento
    def speedUp(self):
        speedUpString = "SpeedUp in progres: "
        for i in range(self.velocidad):
            speedUpString += "🛫"

        for i in range(10):
            speedUpString += "🛫"
            print(speedUpString)
            time.sleep(1 - (i * 0.05))

        self.addSpeed(10)

    def addSpeed(self, velocidad):
        self.velocidad += velocidad
        return self.velocidad

    def downSpeed(self, velocidad):
        self.velocidad -= velocidad
        return self.velocidad

    def currentSpeed(self):
        print("Current Speed 🛫: " + str(self.velocidad))

    def landing(self):
        for i in range(5):
            print("🛬🛬 Landing in Progress 🚷🚷")
            time.sleep(1 - (i * 0.05))

    def shootMissiles(self):
        for i in range(5):
            print("🚀🚀 Missiles ready to fire 🔥🔥🔥")
            time.sleep(1 - (i * 0.05))

    def irrigation(self):
        for i in range(5):
            print("🛩️🛩️ irrigation in process 💧💧💧")
            time.sleep(1 - (i * 0.05))

    def descent(self):
        speedUpString = ""
        for i in range(self.velocidad):
            speedUpString += "🛬"

        for i in range(10):
            speedUpString = speedUpString[:-1]
            print("🛬🛬Descent in Progress🛬🛬: " + speedUpString)
            time.sleep(1 - (abs(i) * 0.05))

        self.downSpeed(10)

    # Getters and Setters
    def getMarca(self):
        return self.marca

    def setMarca(self, marca):
        self.marca = marca

    def getTipo(self):
        return self.tipo

    def setTipo(self, tipo):
        self.tipo = tipo

    def getVelocidad(self):
        return self.velocidad

    def setVelocidad(self, velocidad):
        self.velocidad = velocidad

File: ruby/pgms_ruby/06-pgmAvion/Avion.rb
require 'time'

class Avion
  def initialize
    @marca = nil
    @tipo = nil
    @velocidad = 0
  end

  def to_s
    "Clase Abstracta--> Avion: #{@marca} #{@tipo} "
  end

  # Abstract Behavior
  def startingEngine
    raise NotImplementedError, "Subclass must implement abstract method"
  end

  def stopingEngine
    raise NotImplementedError, "Subclass must implement abstract method"
  end

  # Definir comportamiento
  def speedUp
    speedUpString = "SpeedUp in progres: "
    @velocidad.times do
      speedUpString += "🛫"
    end

    10.times do |i|
      speedUpString += "🛫"
      puts speedUpString
      sleep(1 - (i * 0.05))
    end

    addSpeed(10)
  end

  def addSpeed(velocidad)
    @velocidad += velocidad
    return @velocidad
  end

  def downSpeed(velocidad)
    @velocidad -= velocidad
    return @velocidad
  end

  def currentSpeed
    puts "Current Speed 🛫: #{@velocidad}"
  end

  def landing
     5.times do |i|
      puts "🛬🛬 Landing in Progress 🚷🚷"
      sleep(1 - (i * 0.05))
    end
  end

  def shootMissiles
    5.times do |i|
      puts "🚀🚀 Missiles ready to fire 🔥🔥🔥"
      sleep(1 - (i * 0.05))
    end
  end

  def irrigation
    5.times do |i|
      puts "🛩️🛩️ irrigation in process 💧💧💧"
      sleep(1 - (i * 0.05))
    end
  end

  def descent
    speedUpString = ""
    @velocidad.times do
      speedUpString += "🛬"
    end

    10.times do |i|
      speedUpString = speedUpString[0..-2]
      puts "🛬🛬Descent in Progress🛬🛬: " + speedUpString
      sleep(1 - (@velocidad + i * 0.05))
    end

    downSpeed(10)
  end

  # Getters and Setters
  def getMarca
    return @marca
  end

  def setMarca(marca)
    @marca = marca
  end

  def getTipo
    return @tipo
  end

  def setTipo(tipo)
    @tipo = tipo
  end

  def getVelocidad
    return @velocidad
  end

  def setVelocidad(velocidad)
    @velocidad = velocidad
  end
end